src/operator/index.rs

```diff
@@ -0,0 +1,209 @@
+    pub fn new() -> Self {
+        Self { _type: PhantomData }
+    }
+}
+
+impl<K, V, W, CI, CO, Z> Default for Index<K, V, W, CI, CO, Z> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<K, V, W, CI, CO, Z> Operator for Index<K, V, W, CI, CO, Z>
+where
+    K: 'static,
+    V: 'static,
+    W: 'static,
+    CI: 'static,
+    CO: 'static,
+    Z: 'static,
+{
+    fn name(&self) -> Cow<'static, str> {
+        Cow::from("Index")
+    }
+    fn clock_start(&mut self, _scope: Scope) {}
+    fn clock_end(&mut self, _scope: Scope) {}
+}
+
+impl<K, V, W, CI, CO, Z> UnaryOperator<CI, CO> for Index<K, V, W, CI, CO, Z>
+where
+    K: KeyProperties,
+    V: KeyProperties,
+    W: ZRingValue,
+    CI: IntoIterator<Item = ((K, V), W)> + 'static,
+    for<'a> &'a CI: IntoIterator,
+    for<'a> <&'a CI as IntoIterator>::Item: RefPair<'a, (K, V), W>,
+    CO: FiniteMap<K, Z> + Default,
+    Z: ZSet<V, W>,
+{
+    fn eval(&mut self, i: &CI) -> CO {
+        let mut res = CO::default();
+        for pair in i.into_iter() {
+            let ((k, v), w) = pair.into_refs();
+            res.update(k, |val| val.increment(v, w.clone()));
+        }
+        res
+    }
+
+    fn eval_owned(&mut self, i: CI) -> CO {
+        let mut res = CO::default();
+        for ((k, v), w) in i.into_iter() {
+            res.update_owned(k, |val| val.increment_owned(v, w.clone()));
+        }
+        res
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::{
+        algebra::{FiniteHashMap, FiniteMap},
+        circuit::Root,
+        finite_map,
+        operator::Generator,
+    };
+
+    #[test]
+    fn index_sequence() {
+        let root = Root::build(move |circuit| {
+            let mut inputs = vec![
+                vec![ ((1, "a"), 1)
+                    , ((1, "b"), 1)
+                    , ((2, "a"), 1)
+                    , ((2, "c"), 1)
+                    , ((1, "a"), 2)
+                    , ((1, "b"), -1)
+                ],
+                vec![ ((1, "d"), 1)
+                    , ((1, "e"), 1)
+                    , ((2, "a"), -1)
+                    , ((3, "a"), 2)
+                ],
+            ].into_iter();
+            let mut outputs = vec![
+                finite_map!{ 1 => finite_map!{"a" => 3}, 2 => finite_map!{"a" => 1, "c" => 1}},
+                finite_map!{ 1 => finite_map!{"a" => 3, "d" => 1, "e" => 1}, 2 => finite_map!{"c" => 1}, 3 => finite_map!{"a" => 2}},
+            ].into_iter();
+            circuit.add_source(Generator::new(move || inputs.next().unwrap() ))
+                   .index()
+                   .integrate()
+                   .inspect(move |fm: &FiniteHashMap<_, _>| assert_eq!(fm, &outputs.next().unwrap()));
+        })
+        .unwrap();
+
+        for _ in 0..2 {
+            root.step().unwrap();
+        }
+    }
+
+    #[test]
+    fn index_zset() {
+        let root = Root::build(move |circuit| {
+            let mut inputs = vec![
+                finite_map!{
```

Comment:
how about a test where you really partition the zset using some key function?

Reply:
Yes, would be nice, but we'll need a `Map` operator first.  I'll add a couple more operators (Map, Join) and write more tests using them.

Reply:
Isn't Apply a map?

Reply:
Yes, but not a very good one.  I think we can do better performance-wise.